runtime/lua/vim/_system.lua

```diff
@@ -0,0 +1,360 @@
+    stdin:write('', function()
+      stdin:shutdown(function()
+        if stdin then
+          stdin:close()
+        end
+      end)
+    end)
+  end
+end
+
+--- @return boolean
+function SystemObj:is_closing()
+  local handle = self._state.handle
+  return handle == nil or handle:is_closing()
+end
+
+--- @param spec string[]|SystemSpec
+--- @return SystemSpec
+local function process_spec(spec)
+  if spec[1] then
+    local cmd = {} --- @type string[]
+    for _, p in ipairs(spec) do
+      cmd[#cmd + 1] = p
+    end
+    spec = vim.deepcopy(spec)
+    spec.cmd = cmd
+  end
+
+  return spec
+end
+
+---@private
+---@param output function|'false'
+---@return uv_stream_t?
+---@return function? Handler
+local function setup_output(output)
+  if output == nil then
+    return assert(uv.new_pipe(false)), nil
+  end
+
+  if type(output) == 'function' then
+    return assert(uv.new_pipe(false)), output
+  end
+
+  assert(output == false)
+  return nil, nil
+end
+
+---@private
+---@param input string|string[]|true|nil
+---@return uv_stream_t?
+---@return string|string[]?
+local function setup_input(input)
+  if not input then
+    return
+  end
+
+  local towrite --- @type string|string[]?
+  if type(input) == 'string' or type(input) == 'table' then
+    towrite = input
+  end
+
+  return assert(uv.new_pipe(false)), towrite
+end
+
+--- @param cmd string|string[]
+--- @param args? string[]
+--- @return string Command
+--- @return string[]? Arguments
+local function setup_cmd(cmd, args)
+  if type(cmd) == 'string' then
+    cmd = { cmd }
+    if args then
+      vim.list_extend(cmd, args)
+    end
+  end
+
+  return cmd[1], vim.list_slice(cmd, 2)
+end
+
+--- uv.spawn will completely overwrite the environment
+--- when we just want to modify the existing one, so if env is provided
+--- make sure to prepopulate it with the current env.
+--- @param env table<string,string|number>
+--- @return string[]?
+local function setup_env(env)
+  if not env then
+    return
+  end
+
+  --- @type table<string,string>
+  local env0 = vim.tbl_extend('force', vim.fn.environ(), env)
+
+  local renv = {} --- @type string[]
+  for k, v in pairs(env0) do
+    renv[#renv + 1] = string.format('%s=%s', k, tostring(v))
+  end
+
+  return renv
+end
+
+--- @param stream uv_stream_t
+--- @param bucket string[]
+--- @return fun(err: string?, data: string?)
+local function default_handler(stream, bucket)
+  return function(err, data)
+    if err then
+      error(err)
+    end
+    if data ~= nil then
+      bucket[#bucket + 1] = data
+    else
+      stream:read_stop()
+      stream:close()
+    end
+  end
+end
+
+local M = {}
+
+--- @param cmd string
+--- @param opts uv.aliases.spawn_options
+--- @param on_exit fun(code: integer, signal: integer)
+--- @param on_error fun()
+--- @return uv_process_t, integer
+local function spawn(cmd, opts, on_exit, on_error)
+  local handle, pid_or_err = uv.spawn(cmd, opts, on_exit)
```

Comment:
If this doesn't use the internals of `jobstart` then we probably need special handling similar to `:help jobstart-env`. In particular, unsetting `$NVIM_LISTEN_ADDRESS` is important. The others could be a backlog item (i.e. not a blocker for this PR).

Reply:
What problem does this avoid? None of the users of `uv.spawn()` I've found handle this. Also `NVIM_LISTEN_ADDRESS` is deprecated?

Reply:
> None of the users of `uv.spawn()` I've found handle this

This is something Nvim should handle itself. It's for robustness: 

- we don't want anyone accidentally taking a dependency on `$NVIM_LISTEN_ADDRESS` (which, yes, is deprecated, and explicitly should not be passed to child processes to further strengthen the contract and avoid confusion)
- `$NVIM` should always be set on child processes, else it serves no purpose